File: TrappingRainWater/TrappingRainWater.cpp
#include <bits/stdc++.h>
#include "../crio/cpp/io/FastIO.hpp"
#include "../crio/cpp/io/ReadMatrix.hpp"
using namespace std;

class TrappingRainWater {
public:
	// Implement your solution by completing the below function`	
	int trap(vector<int>& height) {
	    int ans = 0;

	    return ans;
	}
};

int main() {
	FastIO();
	int n;
	cin >> n;
	vector<int> heights;
	ReadMatrix<int>().OneDMatrix(n, heights);
	int result = TrappingRainWater().trap(heights);
	cout << result;
	return 0;
}

File: TrappingRainWater/TrappingRainWater.py
def trap(arr, n):

    result = 0
    left_max = 0
    right_max = 0
    lo = 0
    hi = n-1
       
    while(lo <= hi):  
      
        if(arr[lo] < arr[hi]): 
          
            if(arr[lo] > left_max): 
                left_max = arr[lo] 
            else: 
                result += left_max - arr[lo] 
            lo+= 1
          
        else: 
          
            if(arr[hi] > right_max): 
                right_max = arr[hi] 
            else: 
                result += right_max - arr[hi] 
            hi-= 1
          
    return result

if __name__ == '__main__':
    n = int(input())
    heights = []
    if (n != 0):
        heights = input().split()
        heights = [int(i) for i in heights]
    result = trap(heights, len(heights))
    print(result)